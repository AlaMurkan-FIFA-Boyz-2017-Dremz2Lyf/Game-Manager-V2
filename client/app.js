// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import promise from 'redux-promise';

// Components
import Home from './components/home';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(promise)
));

ReactDOM.render(
  <Provider store={store}>
    <Home/>
  </Provider>,
  document.getElementById('app')
);
