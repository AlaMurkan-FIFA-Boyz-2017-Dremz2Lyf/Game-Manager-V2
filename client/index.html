<html>
<head>
  <title>Game Manager</title>
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans|Oswald|PT+Sans">
</head>
<body>
  <div id="app"></div>
</body>
</html>
