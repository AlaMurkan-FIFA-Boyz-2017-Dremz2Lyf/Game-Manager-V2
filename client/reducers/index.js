import { combineReducers } from 'redux';
import data from './data_reducer';
import isError from './isError';
import isLoading from './isLoading';

const rootReducer = combineReducers({
  data,
  isError,
  isLoading
});

export default rootReducer;
