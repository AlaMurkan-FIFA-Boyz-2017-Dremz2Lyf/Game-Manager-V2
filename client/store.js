
// Make our store exportable for using in tests/and production

import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from './reducers/index.js';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(rootReducer, undefined, composeEnhancers(
  applyMiddleware(thunk)
));
