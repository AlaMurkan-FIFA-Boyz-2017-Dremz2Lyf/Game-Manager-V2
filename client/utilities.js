/**
  * normalize
  * @param {array} list - an array of objects to be normalized into an object
*/
export const normalize = (list) => {

  return list.reduce((normalized, item) => {
    normalized[item.id] = item;
    return normalized;
  }, {});

};

/**
  * applyAll - apply all uses a bit of tricky programming here to return a new state adding to only the specific substate we need.
  * @param {object} state - the current state
  * @param {object} action - the action coming through, it's payload must have a stateKey and the data corresponding to that statekey.
*/
export const applyAll = (state, action) => {
  const { stateKey, data } = action.payload;

  return {...state, [stateKey]: data};
};

/**
  * percentPlayed - basic math here.
  * @param {number} gamesPlayed - pretty self explanatory
  * @param {number} totalGames - pretty self explanatory
*/
export const percentPlayed = (gamesPlayed, totalGames) => {
  return Math.floor((gamesPlayed / totalGames) * 100);
};

/**
  * getValidationState - will give us the correct bootstrap class for UI responsiveness
  * @param {bool} valid - passed from redux-form, lets us know if the form is valid
  * @param {bool} pristine - from redux-form, lets us know if the form has been touched & edited
*/
export const getValidationState = (valid, pristine) => pristine ? null : (valid ? 'success' : 'error' );

/**
  * required - another validation check for redux-form
  * @param {string} value - from redux-form, returns an alert if the user has not entered required info. Also deals with the number 0 as that is falsey value. 
*/
export const required = value => value === 0 || value ? undefined : 'Required';

/**
  * possValidation - validation specific to the possession field
  * @param {string} value - the input from the form
  * @param {string} allValues - all the inputs from the form
*/
export const possValidation = (value, allValues) => {
  const { p1Poss, p2Poss } = allValues;

  if (!p1Poss && !p2Poss) {
    return 'Possession is required.';
  } else if (Number(p1Poss) + Number(p2Poss) !== 100) {
    return 'Possession should add up to 100%!';
  }
  return undefined;
};

/**
  * helpMessage - renders the correct help message based on the field
  * @param {string} field
*/
export const helpMessage = (field) => {
  let message;

  if (field === 'Score') {
    message = `${field} is required! I mean.. come on!`;
  } else {
    message = `${field} will default to 0 if not given.`;
  }
  return message;
};

/**
  * buildTable
  * @param {object} games - a normalized games object to build the table from
  * @param {object} players - another normalized object to reference the player names.
*/
export const buildTable = (games, players) => {

  let trackedStats = {
    wins: 0, losses: 0, draws: 0, goalsFor: 0, goalsAgainst: 0, shots: 0, onGoal: 0, reds: 0, yellows: 0, poss: null, points: 0, goalDiff: 0
  };

  return Object.values(games).reduce((table, game) => {

    let { p1, p2 } = game;

    table[p1] = table[p1] ? table[p1] : { ...trackedStats, id: p1, username: players[p1].username};
    table[p2] = table[p2] ? table[p2] : { ...trackedStats, id: p2, username: players[p2].username};

    return applyGame(table, game);
  }, {});
};


// I extracted this out of the reduce in build table because of all the action that needs to happen.
// There might be a more programmatic way to do this. Potentially mapping the game objects from the database into nested objects for each player in the game.... that would, in theory, allow us to just loop through each player's object from the game and apply them to the correct player.

/**
  * applyGame - used in buildTable to apply the game to the table
  * @param {object} table - the current table
  * @param {object} game - the current game to be applied
*/
export const applyGame = (table, game) => {

  let { p1, p2, p1Score, p2Score, p1Shots, p2Shots, p1Poss, p2Poss, p1OnGoal, p2OnGoal, p1Reds, p1Yellows, p2Reds, p2Yellows } = game;


  let player1 = table[p1];
  let player2 = table[p2];

  if (p1Score && p2Score) {

    p1Score === p2Score ? (
      player1.draws++,
      player2.draws++
    ) : p1Score > p2Score ? (
      player1.wins++,
      player2.losses++
    ) : (
      player2.wins++,
      player1.losses++
    );
    player1.goalsFor += p1Score;
    player2.goalsFor += p2Score;
    player1.goalsAgainst += p2Score;
    player2.goalsAgainst += p1Score;
    player1.shots += p1Shots;
    player2.shots += p2Shots;
    player1.onGoal += p1OnGoal;
    player2.onGoal += p2OnGoal;
    player1.reds += p1Reds;
    player2.reds += p2Reds;
    player1.yellows += p1Yellows;
    player2.yellows += p2Yellows;
    player1.poss = player1.poss === null ? p1Poss : (player1.poss + p1Poss) / (player1.wins + player1.losses + player1.draws);
    player2.poss = player2.poss === null ? p2Poss : (player1.poss + p1Poss) / (player1.wins + player1.losses + player1.draws);
    player1.points = (player1.wins * 3 + player1.draws);
    player2.points = (player2.wins * 3 + player2.draws);
    player1.goalDiff = (player1.goalsFor - player1.goalsAgainst);
    player2.goalDiff = (player2.goalsFor - player2.goalsAgainst);
  }
  return table;
};


// NOTE: Currently not being used:
  // Need to figure out how to get the react-bootstrap - tooltips to work.
  // Seemingly I need to build a component wrapped in a <th>... but the first attempt didn't work.
export const mapHeaders = (header) => {
  switch (header) {
  case 'P':
    return 'Points';
  case 'W':
    return 'Wins';
  case 'L':
    return 'Losses';
  case 'D':
    return 'Draws';
  case 'GF':
    return 'Goals For';
  case 'GA':
    return 'Goals Against';
  case 'GD':
    return 'Goal Differencial';
  case 'S':
    return 'Shots';
  case 'OT':
    return 'Shots On Target';
  case 'Y':
    return 'Yellows';
  case 'R':
    return 'Reds';
  case 'Po':
    return 'Possession';
  default:
    return 'Player';
  }
};
