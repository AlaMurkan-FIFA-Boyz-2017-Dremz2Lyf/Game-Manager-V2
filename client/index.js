// Dependancies
import React from 'react';
import ReactDOM from 'react-dom';
import { ProviderÂ } from 'react-redux';

// css
import '../styles/styles.css';

// Bring in our router
import Routes from './router';

// Grab our root reducer and our store from their files
import rootReducer from './reducers/index.js';
import { store } from './store';

// Found this... Needed for hot module reloading.
if (module.hot) {
  module.hot.accept();
}

ReactDOM.render(
  <Provider store={store}>
    <Routes />
  </Provider>,
  document.getElementById('app')
);
