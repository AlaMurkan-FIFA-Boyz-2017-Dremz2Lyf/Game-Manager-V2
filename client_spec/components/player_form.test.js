import React from 'react';
import { Provider } from 'react-redux';

import { store } from '../../client/store';
import Connected, { PlayerForm, required, getValidationState, popover } from '../../client/components/player_form';
import { normalize } from '../../client/utilities';

describe('PlayerForm', () => {
  let mounted = mount(
    <Provider store={store}>
      <Connected/>
    </Provider>
  );

  test('should match the snapshot', () => {

    expect(mounted).toMatchSnapshot();
  });

  it('should handle required validation', () => {
    expect(required()).toBe('What? are you no one? A faceless man?');
    expect(required('something')).toBeUndefined();
  });

  it('should have a popover', () => {
    let popWrapper = shallow(popover);
    expect(popWrapper).toMatchSnapshot();
  });

  describe('Prop mocking', () => {
    let submitMock = jest.fn();
    let wrapper = shallow(
      <PlayerForm
        allPlayers={mockData.players}
        handleSubmit={submitMock}
        isPristine={true}
        isValid={false}
        syncErrors={{username: 'Tough luck, someone beat you to it.'}}
      />
    );

    it('should have a handleSubmit prop, called on submit', () => {
      wrapper.find('form').simulate('submit');

      expect(submitMock).toHaveBeenCalled();
    });

    it('should add a HelpBlock if there was an error', () => {

      expect(wrapper.contains('Tough luck, someone beat you to it.')).toBe(false);

      wrapper.setProps({isPristine: false});

      expect(wrapper.contains('Tough luck, someone beat you to it.')).toBe(true);

      wrapper.setProps({isValid: true});

      expect(wrapper.contains('Tough luck, someone beat you to it.')).toBe(false);
    });

    it('notTaken should check against the players if the name is taken.', () => {
      let noError = wrapper.instance().notTaken('scott');
      let taken = wrapper.instance().notTaken('alice');
      expect(noError).toBeUndefined();
      expect(taken).toBe('Tough luck, someone beat you to it.');
    });

  });


});
