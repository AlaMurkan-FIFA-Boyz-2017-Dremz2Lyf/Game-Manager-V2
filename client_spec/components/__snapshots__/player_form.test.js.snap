exports[`Connected PlayerForm should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }>
  <ReduxForm>
    <Connect(Form(Connect(PlayerForm)))
      destroyOnUnmount={true}
      enableReinitialize={false}
      forceUnregisterOnUnmount={false}
      form="playerForm"
      getFormState={[Function]}
      keepDirtyOnReinitialize={false}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldValidate={[Function]}
      touchOnBlur={true}
      touchOnChange={false}>
      <Form(Connect(PlayerForm))
        anyTouched={false}
        array={
          Object {
            "insert": [Function],
            "move": [Function],
            "pop": [Function],
            "push": [Function],
            "remove": [Function],
            "removeAll": [Function],
            "shift": [Function],
            "splice": [Function],
            "swap": [Function],
            "unshift": [Function],
          }
        }
        arrayInsert={[Function]}
        arrayMove={[Function]}
        arrayPop={[Function]}
        arrayPush={[Function]}
        arrayRemove={[Function]}
        arrayRemoveAll={[Function]}
        arrayShift={[Function]}
        arraySplice={[Function]}
        arraySwap={[Function]}
        arrayUnshift={[Function]}
        asyncValidating={false}
        autofill={[Function]}
        blur={[Function]}
        change={[Function]}
        clearAsyncError={[Function]}
        clearSubmit={[Function]}
        destroy={[Function]}
        destroyOnUnmount={true}
        dirty={false}
        dispatch={[Function]}
        enableReinitialize={false}
        focus={[Function]}
        forceUnregisterOnUnmount={false}
        form="playerForm"
        getFormState={[Function]}
        initialize={[Function]}
        initialized={false}
        invalid={true}
        keepDirtyOnReinitialize={false}
        persistentSubmitErrors={false}
        pristine={true}
        pure={true}
        registerField={[Function]}
        registeredFields={
          Object {
            "isTeam": Object {
              "count": 1,
              "name": "isTeam",
              "type": "Field",
            },
            "username": Object {
              "count": 1,
              "name": "username",
              "type": "Field",
            },
          }
        }
        reset={[Function]}
        setSubmitFailed={[Function]}
        setSubmitSucceeded={[Function]}
        shouldAsyncValidate={[Function]}
        shouldValidate={[Function]}
        startAsyncValidation={[Function]}
        startSubmit={[Function]}
        stopAsyncValidation={[Function]}
        stopSubmit={[Function]}
        submit={[Function]}
        submitFailed={false}
        submitSucceeded={false}
        submitting={false}
        syncErrors={
          Object {
            "username": "Required",
          }
        }
        syncWarnings={Object {}}
        touch={[Function]}
        touchOnBlur={true}
        touchOnChange={false}
        unregisterField={[Function]}
        untouch={[Function]}
        updateSyncErrors={[Function]}
        updateSyncWarnings={[Function]}
        valid={false}
        validExceptSubmit={false}
        values={Object {}}>
        <Connect(PlayerForm)
          anyTouched={false}
          array={
            Object {
              "insert": [Function],
              "move": [Function],
              "pop": [Function],
              "push": [Function],
              "remove": [Function],
              "removeAll": [Function],
              "shift": [Function],
              "splice": [Function],
              "swap": [Function],
              "unshift": [Function],
            }
          }
          asyncValidate={[Function]}
          asyncValidating={false}
          autofill={[Function]}
          blur={[Function]}
          change={[Function]}
          clearAsyncError={[Function]}
          clearSubmit={[Function]}
          destroy={[Function]}
          dirty={false}
          dispatch={[Function]}
          form="playerForm"
          handleSubmit={[Function]}
          initialize={[Function]}
          initialized={false}
          invalid={true}
          pristine={true}
          pure={true}
          reset={[Function]}
          submit={[Function]}
          submitFailed={false}
          submitSucceeded={false}
          submitting={false}
          touch={[Function]}
          untouch={[Function]}
          valid={false}>
          <PlayerForm
            allPlayers={Object {}}
            anyTouched={false}
            array={
              Object {
                "insert": [Function],
                "move": [Function],
                "pop": [Function],
                "push": [Function],
                "remove": [Function],
                "removeAll": [Function],
                "shift": [Function],
                "splice": [Function],
                "swap": [Function],
                "unshift": [Function],
              }
            }
            asyncValidate={[Function]}
            asyncValidating={false}
            autofill={[Function]}
            blur={[Function]}
            change={[Function]}
            clearAsyncError={[Function]}
            clearSubmit={[Function]}
            destroy={[Function]}
            dirty={false}
            dispatch={[Function]}
            form="playerForm"
            handleSubmit={[Function]}
            initialize={[Function]}
            initialized={false}
            invalid={true}
            isPristine={true}
            isValid={false}
            pristine={true}
            pure={true}
            reset={[Function]}
            submit={[Function]}
            submitFailed={false}
            submitSucceeded={false}
            submitting={false}
            syncErrors={
              Object {
                "username": "Required",
              }
            }
            touch={[Function]}
            untouch={[Function]}
            valid={false}>
            <form
              onSubmit={[Function]}>
              <div>
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}>
                  <label
                    className="control-label">
                    Username
                  </label>
                </ControlLabel>
                <InputGroup
                  bsClass="input-group">
                  <span
                    className="input-group">
                    <OverlayTrigger
                      defaultOverlayShown={false}
                      overlay={
                        <Popover
                          bsClass="popover"
                          id="popover-trigger-hover-focus"
                          placement="right">
                          Select this to make a doubles team.
                        </Popover>
                      }
                      placement="bottom"
                      trigger={
                        Array [
                          "hover",
                          "focus",
                        ]
                      }>
                      <InputGroupAddon
                        aria-describedby="popover-trigger-hover-focus"
                        bsClass="input-group-addon"
                        onBlur={[Function]}
                        onClick={null}
                        onFocus={[Function]}
                        onMouseOut={[Function]}
                        onMouseOver={[Function]}>
                        <span
                          aria-describedby="popover-trigger-hover-focus"
                          className="input-group-addon"
                          onBlur={[Function]}
                          onClick={null}
                          onFocus={[Function]}
                          onMouseOut={[Function]}
                          onMouseOver={[Function]}>
                          <Field
                            autoComplete="false"
                            component="input"
                            label="Doubles Team?"
                            name="isTeam"
                            type="checkbox">
                            <Connect(ConnectedField)
                              _reduxForm={
                                Object {
                                  "anyTouched": false,
                                  "array": Object {
                                    "insert": [Function],
                                    "move": [Function],
                                    "pop": [Function],
                                    "push": [Function],
                                    "remove": [Function],
                                    "removeAll": [Function],
                                    "shift": [Function],
                                    "splice": [Function],
                                    "swap": [Function],
                                    "unshift": [Function],
                                  },
                                  "arrayInsert": [Function],
                                  "arrayMove": [Function],
                                  "arrayPop": [Function],
                                  "arrayPush": [Function],
                                  "arrayRemove": [Function],
                                  "arrayRemoveAll": [Function],
                                  "arrayShift": [Function],
                                  "arraySplice": [Function],
                                  "arraySwap": [Function],
                                  "arrayUnshift": [Function],
                                  "asyncErrors": undefined,
                                  "asyncValidate": [Function],
                                  "asyncValidating": false,
                                  "autofill": [Function],
                                  "blur": [Function],
                                  "change": [Function],
                                  "clearAsyncError": [Function],
                                  "clearSubmit": [Function],
                                  "destroy": [Function],
                                  "destroyOnUnmount": true,
                                  "dirty": false,
                                  "dispatch": [Function],
                                  "enableReinitialize": false,
                                  "error": undefined,
                                  "focus": [Function],
                                  "forceUnregisterOnUnmount": false,
                                  "form": "playerForm",
                                  "getFormState": [Function],
                                  "getValues": [Function],
                                  "initialValues": undefined,
                                  "initialize": [Function],
                                  "initialized": false,
                                  "invalid": false,
                                  "keepDirtyOnReinitialize": false,
                                  "persistentSubmitErrors": false,
                                  "pristine": true,
                                  "pure": true,
                                  "register": [Function],
                                  "registerField": [Function],
                                  "registerInnerOnSubmit": [Function],
                                  "registeredFields": undefined,
                                  "reset": [Function],
                                  "sectionPrefix": undefined,
                                  "setSubmitFailed": [Function],
                                  "setSubmitSucceeded": [Function],
                                  "shouldAsyncValidate": [Function],
                                  "shouldValidate": [Function],
                                  "startAsyncValidation": [Function],
                                  "startSubmit": [Function],
                                  "stopAsyncValidation": [Function],
                                  "stopSubmit": [Function],
                                  "submit": [Function],
                                  "submitFailed": false,
                                  "submitSucceeded": false,
                                  "submitting": false,
                                  "syncErrors": Object {},
                                  "syncWarnings": Object {},
                                  "touch": [Function],
                                  "touchOnBlur": true,
                                  "touchOnChange": false,
                                  "triggerSubmit": undefined,
                                  "unregister": [Function],
                                  "unregisterField": [Function],
                                  "untouch": [Function],
                                  "updateSyncErrors": [Function],
                                  "updateSyncWarnings": [Function],
                                  "valid": true,
                                  "validExceptSubmit": true,
                                  "values": Object {},
                                  "warning": undefined,
                                }
                              }
                              autoComplete="false"
                              component="input"
                              label="Doubles Team?"
                              name="isTeam"
                              normalize={[Function]}
                              type="checkbox">
                              <ConnectedField
                                _reduxForm={
                                  Object {
                                    "anyTouched": false,
                                    "array": Object {
                                      "insert": [Function],
                                      "move": [Function],
                                      "pop": [Function],
                                      "push": [Function],
                                      "remove": [Function],
                                      "removeAll": [Function],
                                      "shift": [Function],
                                      "splice": [Function],
                                      "swap": [Function],
                                      "unshift": [Function],
                                    },
                                    "arrayInsert": [Function],
                                    "arrayMove": [Function],
                                    "arrayPop": [Function],
                                    "arrayPush": [Function],
                                    "arrayRemove": [Function],
                                    "arrayRemoveAll": [Function],
                                    "arrayShift": [Function],
                                    "arraySplice": [Function],
                                    "arraySwap": [Function],
                                    "arrayUnshift": [Function],
                                    "asyncErrors": undefined,
                                    "asyncValidate": [Function],
                                    "asyncValidating": false,
                                    "autofill": [Function],
                                    "blur": [Function],
                                    "change": [Function],
                                    "clearAsyncError": [Function],
                                    "clearSubmit": [Function],
                                    "destroy": [Function],
                                    "destroyOnUnmount": true,
                                    "dirty": false,
                                    "dispatch": [Function],
                                    "enableReinitialize": false,
                                    "error": undefined,
                                    "focus": [Function],
                                    "forceUnregisterOnUnmount": false,
                                    "form": "playerForm",
                                    "getFormState": [Function],
                                    "getValues": [Function],
                                    "initialValues": undefined,
                                    "initialize": [Function],
                                    "initialized": false,
                                    "invalid": false,
                                    "keepDirtyOnReinitialize": false,
                                    "persistentSubmitErrors": false,
                                    "pristine": true,
                                    "pure": true,
                                    "register": [Function],
                                    "registerField": [Function],
                                    "registerInnerOnSubmit": [Function],
                                    "registeredFields": undefined,
                                    "reset": [Function],
                                    "sectionPrefix": undefined,
                                    "setSubmitFailed": [Function],
                                    "setSubmitSucceeded": [Function],
                                    "shouldAsyncValidate": [Function],
                                    "shouldValidate": [Function],
                                    "startAsyncValidation": [Function],
                                    "startSubmit": [Function],
                                    "stopAsyncValidation": [Function],
                                    "stopSubmit": [Function],
                                    "submit": [Function],
                                    "submitFailed": false,
                                    "submitSucceeded": false,
                                    "submitting": false,
                                    "syncErrors": Object {},
                                    "syncWarnings": Object {},
                                    "touch": [Function],
                                    "touchOnBlur": true,
                                    "touchOnChange": false,
                                    "triggerSubmit": undefined,
                                    "unregister": [Function],
                                    "unregisterField": [Function],
                                    "untouch": [Function],
                                    "updateSyncErrors": [Function],
                                    "updateSyncWarnings": [Function],
                                    "valid": true,
                                    "validExceptSubmit": true,
                                    "values": Object {},
                                    "warning": undefined,
                                  }
                                }
                                asyncValidating={false}
                                autoComplete="false"
                                component="input"
                                dirty={false}
                                dispatch={[Function]}
                                label="Doubles Team?"
                                name="isTeam"
                                normalize={[Function]}
                                pristine={true}
                                type="checkbox">
                                <input
                                  autoComplete="false"
                                  checked={false}
                                  label="Doubles Team?"
                                  name="isTeam"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onDragStart={[Function]}
                                  onDrop={[Function]}
                                  onFocus={[Function]}
                                  type="checkbox"
                                  value="" />
                              </ConnectedField>
                            </Connect(ConnectedField)>
                          </Field>
                        </span>
                      </InputGroupAddon>
                    </OverlayTrigger>
                    <Field
                      component={[Function]}
                      label="Username"
                      name="username"
                      type="text"
                      validate={
                        Array [
                          [Function],
                          [Function],
                        ]
                      }>
                      <Connect(ConnectedField)
                        _reduxForm={
                          Object {
                            "anyTouched": false,
                            "array": Object {
                              "insert": [Function],
                              "move": [Function],
                              "pop": [Function],
                              "push": [Function],
                              "remove": [Function],
                              "removeAll": [Function],
                              "shift": [Function],
                              "splice": [Function],
                              "swap": [Function],
                              "unshift": [Function],
                            },
                            "arrayInsert": [Function],
                            "arrayMove": [Function],
                            "arrayPop": [Function],
                            "arrayPush": [Function],
                            "arrayRemove": [Function],
                            "arrayRemoveAll": [Function],
                            "arrayShift": [Function],
                            "arraySplice": [Function],
                            "arraySwap": [Function],
                            "arrayUnshift": [Function],
                            "asyncErrors": undefined,
                            "asyncValidate": [Function],
                            "asyncValidating": false,
                            "autofill": [Function],
                            "blur": [Function],
                            "change": [Function],
                            "clearAsyncError": [Function],
                            "clearSubmit": [Function],
                            "destroy": [Function],
                            "destroyOnUnmount": true,
                            "dirty": false,
                            "dispatch": [Function],
                            "enableReinitialize": false,
                            "error": undefined,
                            "focus": [Function],
                            "forceUnregisterOnUnmount": false,
                            "form": "playerForm",
                            "getFormState": [Function],
                            "getValues": [Function],
                            "initialValues": undefined,
                            "initialize": [Function],
                            "initialized": false,
                            "invalid": true,
                            "keepDirtyOnReinitialize": false,
                            "persistentSubmitErrors": false,
                            "pristine": true,
                            "pure": true,
                            "register": [Function],
                            "registerField": [Function],
                            "registerInnerOnSubmit": [Function],
                            "registeredFields": Object {
                              "isTeam": Object {
                                "count": 1,
                                "name": "isTeam",
                                "type": "Field",
                              },
                              "username": Object {
                                "count": 1,
                                "name": "username",
                                "type": "Field",
                              },
                            },
                            "reset": [Function],
                            "sectionPrefix": undefined,
                            "setSubmitFailed": [Function],
                            "setSubmitSucceeded": [Function],
                            "shouldAsyncValidate": [Function],
                            "shouldValidate": [Function],
                            "startAsyncValidation": [Function],
                            "startSubmit": [Function],
                            "stopAsyncValidation": [Function],
                            "stopSubmit": [Function],
                            "submit": [Function],
                            "submitFailed": false,
                            "submitSucceeded": false,
                            "submitting": false,
                            "syncErrors": Object {
                              "username": "Required",
                            },
                            "syncWarnings": Object {},
                            "touch": [Function],
                            "touchOnBlur": true,
                            "touchOnChange": false,
                            "triggerSubmit": undefined,
                            "unregister": [Function],
                            "unregisterField": [Function],
                            "untouch": [Function],
                            "updateSyncErrors": [Function],
                            "updateSyncWarnings": [Function],
                            "valid": false,
                            "validExceptSubmit": false,
                            "values": Object {},
                            "warning": undefined,
                          }
                        }
                        component={[Function]}
                        label="Username"
                        name="username"
                        normalize={[Function]}
                        type="text"
                        validate={
                          Array [
                            [Function],
                            [Function],
                          ]
                        }>
                        <ConnectedField
                          _reduxForm={
                            Object {
                              "anyTouched": false,
                              "array": Object {
                                "insert": [Function],
                                "move": [Function],
                                "pop": [Function],
                                "push": [Function],
                                "remove": [Function],
                                "removeAll": [Function],
                                "shift": [Function],
                                "splice": [Function],
                                "swap": [Function],
                                "unshift": [Function],
                              },
                              "arrayInsert": [Function],
                              "arrayMove": [Function],
                              "arrayPop": [Function],
                              "arrayPush": [Function],
                              "arrayRemove": [Function],
                              "arrayRemoveAll": [Function],
                              "arrayShift": [Function],
                              "arraySplice": [Function],
                              "arraySwap": [Function],
                              "arrayUnshift": [Function],
                              "asyncErrors": undefined,
                              "asyncValidate": [Function],
                              "asyncValidating": false,
                              "autofill": [Function],
                              "blur": [Function],
                              "change": [Function],
                              "clearAsyncError": [Function],
                              "clearSubmit": [Function],
                              "destroy": [Function],
                              "destroyOnUnmount": true,
                              "dirty": false,
                              "dispatch": [Function],
                              "enableReinitialize": false,
                              "error": undefined,
                              "focus": [Function],
                              "forceUnregisterOnUnmount": false,
                              "form": "playerForm",
                              "getFormState": [Function],
                              "getValues": [Function],
                              "initialValues": undefined,
                              "initialize": [Function],
                              "initialized": false,
                              "invalid": true,
                              "keepDirtyOnReinitialize": false,
                              "persistentSubmitErrors": false,
                              "pristine": true,
                              "pure": true,
                              "register": [Function],
                              "registerField": [Function],
                              "registerInnerOnSubmit": [Function],
                              "registeredFields": Object {
                                "isTeam": Object {
                                  "count": 1,
                                  "name": "isTeam",
                                  "type": "Field",
                                },
                                "username": Object {
                                  "count": 1,
                                  "name": "username",
                                  "type": "Field",
                                },
                              },
                              "reset": [Function],
                              "sectionPrefix": undefined,
                              "setSubmitFailed": [Function],
                              "setSubmitSucceeded": [Function],
                              "shouldAsyncValidate": [Function],
                              "shouldValidate": [Function],
                              "startAsyncValidation": [Function],
                              "startSubmit": [Function],
                              "stopAsyncValidation": [Function],
                              "stopSubmit": [Function],
                              "submit": [Function],
                              "submitFailed": false,
                              "submitSucceeded": false,
                              "submitting": false,
                              "syncErrors": Object {
                                "username": "Required",
                              },
                              "syncWarnings": Object {},
                              "touch": [Function],
                              "touchOnBlur": true,
                              "touchOnChange": false,
                              "triggerSubmit": undefined,
                              "unregister": [Function],
                              "unregisterField": [Function],
                              "untouch": [Function],
                              "updateSyncErrors": [Function],
                              "updateSyncWarnings": [Function],
                              "valid": false,
                              "validExceptSubmit": false,
                              "values": Object {},
                              "warning": undefined,
                            }
                          }
                          asyncValidating={false}
                          component={[Function]}
                          dirty={false}
                          dispatch={[Function]}
                          label="Username"
                          name="username"
                          normalize={[Function]}
                          pristine={true}
                          syncError="Required"
                          type="text"
                          validate={
                            Array [
                              [Function],
                              [Function],
                            ]
                          }>
                          <Component
                            input={
                              Object {
                                "name": "username",
                                "onBlur": [Function],
                                "onChange": [Function],
                                "onDragStart": [Function],
                                "onDrop": [Function],
                                "onFocus": [Function],
                                "value": "",
                              }
                            }
                            label="Username"
                            meta={
                              Object {
                                "active": false,
                                "asyncValidating": false,
                                "autofilled": false,
                                "dirty": false,
                                "dispatch": [Function],
                                "error": "Required",
                                "form": "playerForm",
                                "invalid": true,
                                "pristine": true,
                                "submitFailed": false,
                                "submitting": false,
                                "touched": false,
                                "valid": false,
                                "visited": false,
                                "warning": undefined,
                              }
                            }
                            type="text">
                            <FormGroup
                              bsClass="form-group"
                              validationState={null}>
                              <div
                                className="form-group">
                                <FormControl
                                  bsClass="form-control"
                                  componentClass="input"
                                  name="username"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onDragStart={[Function]}
                                  onDrop={[Function]}
                                  onFocus={[Function]}
                                  placeholder="Username"
                                  type="text"
                                  value="">
                                  <input
                                    className="form-control"
                                    name="username"
                                    onBlur={[Function]}
                                    onChange={[Function]}
                                    onDragStart={[Function]}
                                    onDrop={[Function]}
                                    onFocus={[Function]}
                                    placeholder="Username"
                                    type="text"
                                    value="" />
                                </FormControl>
                              </div>
                            </FormGroup>
                          </Component>
                        </ConnectedField>
                      </Connect(ConnectedField)>
                    </Field>
                    <InputGroupButton
                      bsClass="input-group-btn">
                      <span
                        className="input-group-btn">
                        <Button
                          active={false}
                          block={false}
                          bsClass="btn"
                          bsStyle="default"
                          disabled={true}
                          type="submit">
                          <button
                            className="btn btn-default"
                            disabled={true}
                            type="submit">
                            Add
                          </button>
                        </Button>
                      </span>
                    </InputGroupButton>
                  </span>
                </InputGroup>
                <HelpBlock
                  bsClass="help-block">
                  <span
                    className="help-block" />
                </HelpBlock>
              </div>
            </form>
          </PlayerForm>
        </Connect(PlayerForm)>
      </Form(Connect(PlayerForm))>
    </Connect(Form(Connect(PlayerForm)))>
  </ReduxForm>
</Provider>
`;

exports[`PlayerForm Popover should match the snapshot 1`] = `
<div
  className="popover right"
  id="popover-trigger-hover-focus"
  role="tooltip"
  style={
    Object {
      "display": "block",
      "left": undefined,
      "top": undefined,
    }
  }>
  <div
    className="arrow"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    } />
  <div
    className="popover-content">
    Select this to make a doubles team.
  </div>
</div>
`;
